fggs/indices.py

```diff
@@ -0,0 +1,1293 @@
+"""
+# Algebraic index types
+
+A tensor dimension often corresponds to a random variable, and indices along
+the dimension often correspond to possible values of the variable.  Because the
+variable can be of product or sum type, the range of indices [0,n) is often
+[0,l*m) representing a Cartesian product, [0,l+m) representing a disjoint
+union, or a nesting of these operations.  This algebraic type structure affects
+our computations because our tensors are often (the result of einsum operations
+involving) sparse factors that pack/unpack product/sum types, such as
+eye(35).reshape(35,5,7) or eye(5,7).  We represent these sparse tensors
+compactly for asymptotic speedups.
+
+The key is the following language of "embeddings" (injective mappings) from
+"physical" indices to "virtual" indices:
+
+    Embedding ::= X(numel)                     -- EmbeddingVar
+                | Embedding * ... * Embedding  -- ProductEmbedding
+                | numel + Embedding + numel    -- SumEmbedding
+
+For example, to describe how a 5*7 "physical" matrix stores part of a
+three-dimensional "virtual" tensor, we would use two EmbeddingVars [X(5), Y(7)]
+to build up three embeddings such as [X(5) * Y(7), X(5), Y(7)].  The three
+embeddings determine the shape of the virtual tensor to be [35, 5, 7].  Element
+(i,j) of the matrix corresponds to element (i*7+j,i,j) of the tensor.  Other
+elements of the tensor are presumed to be zero.
```

Comment:
It might be helpful to choose smaller numbers and actually write the index embedding out. Or actually, I feel that writing the inverse out would be even more helpful, like for the example below,

```
[ [ 0,    0,    0, ... ],
  [ m[0], 0,    0, ...],
  [ 0,    m[1], 0, ...],
  ... ]
```